179-create-zilla-plus-iot-markeplace-docs (#182)
aklivity/zilla-docs: src/solutions/how-tos/confluent-cloud/iot-ingest-control.md

---
icon: aky-zilla-plus

> description: Setup a public to your Confluent Cloud private cluster from anywhere on the internet.

Reviewer: Setup a public ?? to your Confluent Cloud...

This probably just needs to be reworded for IoT solution.

tags:
  - Zilla Plus
---

# Confluent Cloud IoT Ingest and Control

[Available in <ZillaPlus/>](https://www.aklivity.io/products/zilla-plus)
{.zilla-plus-badge .hint-container .info}

::: info Estimated time to complete 10-15 minutes.
:::

## Overview

> The [Zilla Plus for Confluent Cloud](https://aws.amazon.com/marketplace/pp/prodview-eblxkinsqbaks) IoT Ingest and Control Broker lets MQTT clients connect.

Reviewer: ```suggestion
The [Zilla Plus for Confluent Cloud](https://aws.amazon.com/marketplace/pp/prodview-eblxkinsqbaks) IoT Ingest and Control Broker lets MQTT clients connect to Confluent Cloud, publish messages and subscribe to topics.
```

| --------------------------- | ------------------------------------------------------------------------------ | ------------ | -------------------------------------------------------------------------------------------- |
| Secrets Manager             | Yes                                                                            | Startup only | [Not reached](https://docs.aws.amazon.com/general/latest/gr/asm.html#limits_secrets-manager) |
| Certificate Manager         | No<br><br>Private key and certificate can be inline in Secrets Manager instead | Startup only | [Not reached](https://docs.aws.amazon.com/general/latest/gr/acm.html#limits_acm)             |
| Private Certificate Manager | No<br><br>Private key and certificate can be inline in Secrets Manager instead | Startup only | [Not reached](https://docs.aws.amazon.com/general/latest/gr/acm-pca.html#limits_acm-pca)     |

Default [AWS Service Quotas](https://docs.aws.amazon.com/general/latest/gr/aws_service_limits.html) are recommended.

## Prerequisites

> Before setting up internet access to your Confluent Cloud Cluster, you will need the following:

Reviewer: "setting up internet access" reads like Secure Public Access.

- an Confluent Cloud Cluster configured for SASL/SCRAM authentication

::: tip
Check out the [Troubleshooting](./../aws-services/troubleshooting.md) guide if you run into any issues.
:::

### Create the Confluent Cloud Cluster

> This creates your Confluent Cloud cluster.

> An Confluent Cloud cluster is needed for access via the internet. The [Confluent Cloud Quickstart](https://docs.confluent.io/cloud/current/get-started/index.html) will walk you through creating one. You can skip this step if you have already created an Confluent Cloud cluster with equivalent configuration. We will use the below resource names to reference the AWS resources needed in this guide.

Reviewer: "access via the internet" reads like Secure Public Access.

:::

Navigate to the VPC Management Console [Security Groups](https://console.aws.amazon.com/vpc/home#securityGroups:) table. Select the `my-zilla-iot-proxy-sg` security group you just created. You will create an inbound rule to allow all traffic inside itself.

- Add Inbound Rule
  - Type: `All Traffic`
  - Source type: `Custom`
  - Source: `my-zilla-iot-proxy-sg`

Add the `my-zilla-iot-proxy-sg` security group to your VPC Endpoint by finding your `my-cc-iot-vpce` from the [Endpoints table](https://console.aws.amazon.com/vpcconsole/home#Endpoints:).

- Select your VPC endpoint
- `Actions` menu > select `Manage Security Groups`
- Select both security groups:
  - `default`
  - `my-zilla-iot-proxy-sg`
- Save the changes

### Create the <ZillaPlus/> proxy IAM security role

> This creates an IAM security role to enable the required AWS services for the <ZillaPlus/> proxies.

Follow the [Create IAM Role](./../aws-services/create-iam-role.md) guide to create an IAM security role with the following parameters:

::: code-tabs

@tab Name

```text:no-line-numbers

> aklivity-zilla-proxy

Reviewer: Seems like this naming is inconsistent with the rest, suggest we unify the naming pattern, so that permissions can be set differently for different secrets used by different deployments more easily.

```suggestion
my-zilla-iot-role
```

```

@tab Policies

```text:no-line-numbers
AWSCertificateManagerReadOnly
```

:::

#### IAM role Inline Policies

This creates an IAM security role to enable the required AWS services for the <ZillaPlus/> proxies.

::: code-tabs

@tab Name

```text:no-line-numbers
CCProxySecretsManagerRead
```

@tab JSON Summary

```json:no-line-numbers
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "VisualEditor0",
      "Effect": "Allow",
      "Action": [
        "secretsmanager:GetSecretValue",
        "secretsmanager:DescribeSecret"
      ],
      "Resource": [
        "arn:aws:secretsmanager:*:*:secret:wildcard.example.aklivity.io-*",
        "arn:aws:secretsmanager:*:*:secret:zilla-cc-access-*"
      ]
    }
  ]
}
```

:::

> ::: info If you used a different secret name for your certificate key.

Reviewer: What if you used a different secret name for `zilla-cc-access...` ?

Replace `wildcard.example.aklivity.io` in the resource regular expression for:

```text:no-line-numbers
CCProxySecretsManagerRead
```

:::

## Subscribe via AWS Marketplace

The [Zilla Plus for Confluent Cloud](https://aws.amazon.com/marketplace/pp/prodview-eblxkinsqbaks) is available through the AWS Marketplace. You can skip this step if you have already subscribed to Zilla Plus for Confluent Cloud via AWS Marketplace.

To get started, visit the Proxy's Marketplace [Product Page](https://aws.amazon.com/marketplace/pp/prodview-eblxkinsqbaks) and `Subscribe` to the offering. You should now see `Zilla Plus for Confluent Cloud` listed in your [AWS Marketplace](https://console.aws.amazon.com/marketplace) subscriptions.

## Create the Server Certificate

> We need a TLS Server Certificate for your custom DNS wildcard domain that can be trusted by a Kafka Client from anywhere.

Reviewer: MQTT client

- From the `Agreement` section > `Actions` menu > select `Launch CloudFormation stack`
- Select the `CloudFormation Template` > `IoT Ingest and Control` fulfillment option
- Make sure you have selected the desired region selected, such as `us-east-1`
- Click `Continue to Launch`
  - Choose the action `Launch CloudFormation`

Click `Launch` to complete the `Create stack` wizard with the following details:

### Step 1. Create Stack

- Prepare template: `Template is ready`
- Specify template: `Amazon S3 URL`
  - Amazon S3 URL: `(auto-filled)`

### Step 2. Specify stack details

::: code-tabs

@tab Stack name

```text:no-line-numbers
my-zilla-iot-proxy
```

:::

Parameters:

- Network Configuration
  - VPC: `my-cc-iot-vpc`
  - Subnets: `my-cc-iot-subnet-public-1a` `my-cc-iot-subnet-public-1b`
- Confluent Cloud Configuration
  - Access Credentials and Bootstrap Server Secret ARN: `<zilla-cc-access secret ARN>` \*1
  - Kafka Topics:
    - messages: `mqtt-messages`
    - retained: `mqtt-retained`
    - sessions: `mqtt-sessions`
- Zilla Plus Configuration
  - Instance count: `2`
  - Instance type: `t3.small` \*2

>   - Role: `aklivity-zilla-proxy`

Reviewer: ```suggestion
  - Role: `my-zilla-iot-role`
```

  4. Replace with your own custom wildcard DNS pattern.
  5. Follow the [Create Key Pair](./../aws-services/create-key-pair.md) guide to create a new key pair to access EC2 instances via SSH.

### Step 3. Configure stack options: `(use defaults)`

### Step 4. Review

Confirm the stack details are correct and `Submit` to start the CloudFormation deploy.

::: info
When your <ZillaPlus/> proxy is ready, the [CloudFormation console](https://console.aws.amazon.com/cloudformation) will show `CREATE_COMPLETE` for the newly created stack.
:::

## Verify <ZillaPlus/> proxy Service

> This checks that the services and networking were properly configured.

Navigate to the [EC2 running instances dashboard.](https://console.aws.amazon.com/ec2/home#Instances:instanceState=running)

::: note Check your selected region
Make sure you have selected the desired region, such as `US East (N. Virginia) us-east-1`.
:::

Select either of the <ZillaPlus/> proxies launched by the CloudFormation template to show the details.

::: info

> They each have an IAM Role name starting with `aklivity-zilla-proxy`.

Reviewer: ```suggestion
They each have an IAM Role name starting with `my-zilla-iot-role`.
```

::: warning
Replace these TLS bootstrap server names accordingly for your own custom wildcard DNS pattern.
:::

### Subscribe to a topic

Publish two messages to the newly created topic via the following producer command:

```bash:no-line-numbers
bin/kafka-console-producer.sh \
  --topic zilla-plus-test \
  --producer.config confluent.properties \
  --broker-list <tls-bootstrap-server-names>
```

A prompt will appear for you to type in the messages:

```output:no-line-numbers
>This is my first event
>This is my second event
```

### Publish to a topic

Read these messages back via the following consumer command:

```bash:no-line-numbers
bin/kafka-console-consumer.sh \
  --topic zilla-plus-test \
  --from-beginning \
  --consumer.config confluent.properties \
  --bootstrap-server <tls-bootstrap-server-names>
```

You should see the `This is my first event` and `This is my second event` messages.

```output:no-line-numbers
This is my first event
This is my second event

> ```

Reviewer: This section needs to be use an MQTT client like `mosquitto` instead of a Kafka client.
